graviti/portex/field.py

```diff
@@ -85,6 +98,134 @@ def rename(self, old_name: str, new_name: str) -> None:
+            i = self._keys.get(key, -1)
+
+        self._fields[i][key] = value  # type: ignore[index]
+
+    def __delitem__(self, key: Union[int, str]) -> None:
+        if isinstance(key, int):
+            i, key = self._indices[key]
+        else:
+            i = self._keys[key]
+
+        del self._fields[i][key]  # type: ignore[attr-defined]
+
+    def __iter__(self) -> Iterator[str]:
+        for fields in self._fields:
+            yield from fields
+
+    @property
+    def _keys(self) -> Dict[str, int]:
+        return {key: index for index, field in enumerate(self._fields) for key in field}
+
+    @property
+    def _indices(self) -> List[Tuple[int, int]]:
+        return [
+            (index, sub_index)
+            for index, field in enumerate(self._fields)
+            for sub_index in range(len(field))
+        ]
+
+    def insert(self, index: int, name: str, portex_type: PortexType) -> None:
+        """Insert the name and portex_type at the index.
+
+        Arguments:
+            index: The index to insert the field.
+            name: The name of the field to be inserted.
+            portex_type: The portex_type of the field to be inserted.
+
+        Raises:
+            ValueError: When the name already exists in the fields.
+            TypeError: When trying to insert a field into ImmutableFields.
+
+        """
+        if name in self._keys:
+            raise ValueError(f"The '{name}' field already exists")
+
+        i, j = self._indices[index]
+
+        fields = self._fields[i]
+        if not isinstance(fields, MutableFields) and j == 0:
+            i = i - 1
+            fields = self._fields[i]
+            j = len(fields)
+
+        try:
+            fields.insert(j, name, portex_type)  # type: ignore[attr-defined]
+        except TypeError:
+            raise TypeError(f"Cannot insert at index {index} due to immutable fields") from None
+
+    def astype(self, name: str, portex_type: PortexType) -> None:
+        """Convert the type of the field with the given name to the new PortexType.
+
+        Arguments:
+            name: The name of the field to convert.
+            portex_type: The new PortexType of the field to convert to.
+
+        Raises:
+            TypeError: When trying to change a field in ImmutableFields.
+
+        """
+        fields = self._fields[self._keys[name]]
+        if not isinstance(fields, MutableFields):
+            raise TypeError("Cannot change immutable fields")
+
+        fields.astype(name, portex_type)
+
+    def rename(self, old_name: str, new_name: str) -> None:
+        """Rename the name of a field.
+
+        Arguments:
+            old_name: The current name of the field to be renamed.
+            new_name: The new name of the field to assign.
+
+        Raises:
+            TypeError: When trying to rename a field in ImmutableFields.
+
+        """
+        fields = self._fields[self._keys[old_name]]
+        if not isinstance(fields, MutableFields):
```

Comment:
do we need to check the type of fields?
can we just call the method and catch the exception?